processing/src/main/java/org/apache/druid/query/context/ResponseContext.java

```diff
@@ -30,93 +33,334 @@
+ * <li>Values passed within a single server. These are tagged with
+ * visibility {@code NONE}.)</li>
+ * </ul>
+ * Second, it performs multiple tasks:
+ * <ul>
+ * <li>Registers the keys to be used in the header. But, since it also holds
+ * internal information, the internal information also needs keys, though the
+ * corresponding values are never serialized.</li>
+ * <li>Gathers information for the query as a whole.</li>
+ * <li>Merges information back up the query tree: from multiple segments,
+ * from multiple servers, etc.</li>
+ * <li>Manages headers size by dropping fields when the header would get too
+ * large.</li>
+ * </ul>
+ *
+ * A result is that the information the context, when inspected by a calling
+ * query, may be incomplete if some of it was previously dropped by the
+ * called query.
+ *
+ * <h4>API</h4>
+ *
+ * The query profile needs to obtain the full, untruncated information. To do this
+ * it piggy-backs on the set operations to obtain the full value. To ensure this
+ * is possible, code that works with standard values should call the set (or add)
+ * functions provided which will do the needed map update.
+  */
 @PublicApi
 public abstract class ResponseContext
 {
   /**
    * The base interface of a response context key.
    * Should be implemented by every context key.
    */
-  public interface BaseKey
+  public interface Key
   {
+    /**
+     * Hack, pure and simple. The symbol "ResponseContext.Key.ETAG" must exist
+     * to avoid changing a line of code where we have no tests, which causes
+     * the build to fail. Remove this once the proper tests are added.
+     *
+     * @see {@link org.apache.druid.server.QueryResource#doPost}
+     */
+    public static final Key ETAG = Keys.ETAG;
+
     @JsonValue
     String getName();
+
+    /**
+     * The phase (header, trailer, none) where this key is emitted.
+     */
+    Visibility getPhase();
+
     /**
-     * Merge function associated with a key: Object (Object oldValue, Object newValue)
+     * Reads a value of this key from a JSON stream. Used by {@link ResponseContextDeserializer}.
+     */
+    Object readValue(JsonParser jp);
+
+    /**
+     * Merges two values of type T.
+     *
+     * This method may modify "oldValue" but must not modify "newValue".
      */
-    BiFunction<Object, Object, Object> getMergeFunction();
+    Object mergeValues(Object oldValue, Object newValue);
+
+    /**
+     * Returns true if this key can be removed to reduce header size when the
+     * header would otherwise be too large.
+     */
+    @JsonIgnore
+    boolean canDrop();
+  }
+
+  /**
+   * Where the key is emitted, if at all. Values in the context can be for internal
+   * use only: for return before the query results (header) or only after query
+   * results (trailer).
+   */
+  public enum Visibility
+  {
+    /**
+     * Keys that are present in both the "X-Druid-Response-Context" header *and* the response context trailer.
+     */
+    HEADER,
```

Comment:
What sort of keys will be in both the header and the trailer? Also, should it be something like `HEADER_AND_TRAILER`?

Reply:
Thanks. I'm sill curious about what keys will be in both the header and the trailer :slightly_smiling_face: Will the next PR need to modify this enum, such as adding a new `Visibility` type?

Reply:
In Gian's query trailer prototype `uncoveredIntervals` `uncoveredIntervalsOverflowed`, `missingSegments`, `cpuConsumed`, and a new `metrics` item are trailer entries. In the query profile work, the profile is also in the trailer only. The final set of values will be `HEADER`, `TRAILER`, `HEADER_AND_TRAILER`, `NONE`. 

Having just written this, it may make more sense to have two Booleans rather than an all-possible-combination enum. What do you think?

Reply:
Like `isVisibleInHeader` and `isVisibleInTrailer`? That sounds reasonable to me. You can do it in your follow-up PR if you want.

Reply:
@jihoonson, went ahead and converted `Visibility` to a Boolean, since I had to make another fix as well. For now, only the header flag appears: the next PR will add the footer flag.